documentation/puppet_server_metrics_performance.markdown

```diff
@@ -0,0 +1,78 @@
+
+HTTP metrics include the total time for the server to handle the request, including waiting for a JRuby instance to become available. Once JRuby borrow time increases, wait time also increases, so once borrow time for *one* type of request increases, wait times for *all* requests increases.
+
+Catalog compilation, which is graphed on the [sample Grafana dashboard][], most commonly increases request times, because there are many points of potential failure or delay in a catalog compilation. Several things could cause catalog compilation lengthen JRuby borrow times.
+
+-   A Puppet code change, such as a faulty or complex new function. The Grafana dashboard should show if functions start taking significantly longer, and the experimental dashboard and [status API][] endpoint also list the lengthiest function calls (showing the top 10 and top 40, respectively) based on aggregate execution times.
+-   Adding many file resources at once.
+
+In cases like these, there might be more efficient ways to author your Puppet code, you might be extending Puppet to the point where you need to add JRubies or compile masters even if you aren't adding more agents.
+
+Slowdowns in PuppetDB can also cause catalog compilations to take more time: if you use exported resources or the `puppetdb_query` function and PuppetDB has a problem, catalog compilation times will increase.
+
+Puppet Server also sends agents' facts and the compiled catalog to PuppetDB during catalog compilation. The [status API][] for the master service reports metrics for these operations under [`http-client-metrics`][HTTP client metrics], and in the Grafana dashboard in the "External HTTP Communications" graph.
+
+Puppet Server also requests facts as HTTP requests while handling a node request, and submits reports via HTTP requests while handling of a report request. If you have an HTTP report processor set up, the Grafana dashboard shows metrics for `Http report processor,` as does the [status API][] endpoint under `http-client-metrics` in the master service, for metric ID `['puppet', 'report', 'http']`. Delays in the report processor are passed on to Puppet Server.
+
+### Memory leaks and usage
+
+A memory leak or increased memory pressure can stress Puppet Server's available resources. In this case, the Java VM will spend more time doing garbage collection, causing the GC time and GC CPU % metrics to increase. These metrics are available in the [developer dashboard][] and [status API][] endpoint, as well as in the `/metrics/v1/mbeans` endpoint.
+
+If you can't identify the source of a memory leak, you might need to add more RAM to your Server or an additional compile master.
```

Comment:
I hesitate making this recommendation - although I understand some people would probably resort to this anyway.  Adding RAM just prolongs the time to failure with an OutOfMemory error and adding another compile master may not help with preventing the other compile masters from OOMing either.

If anything here, it might be worth pointing out the [max-requests-per-instance](https://github.com/puppetlabs/puppetserver/blob/56ee2195b67db4fd865985081357736d779c3183/documentation/config_file_puppetserver.markdown#settings) setting here as this is frequently used as a way to mitigate the effects of a memory leak, avoiding OutOfMemory crashes but with some performance impact.